### ISSUE TITLE
Defined Compound editing on Substance Page

### ISSUE THREAD

**Issue by cmgrulke** [Chemical-Curation/chemcurator_vuejs#55]
**As a** *curator*
**I want** *to be able edit and submit structures on the substance page*
**So that** *the new structures get stored in the database and compound ids*

**Acceptance Criteria:**
- when a defined compound structure is altered in ketcher, a search using the new structure is done to identify if the defined compound exists.
- if the defined compound exists, its CID is displayed in the CID field on the form (otherwise left blank) 
- if a defined compound does exist, but its structure exported from the ketcher window does not match with the one stored by the api, a "Save Structure" button is available and when the save structure button is pressed, the structure is PATCHED
-- if the save is successful, the compound id is displayed in the form.
-- if the save is not successful, the user is notified of the error
- if a defined compound does not exists for the structure in the editor, a "Save Structure" button is available and when save structure button is pressed, the structure is POSTED to the api
-- if the save is successful, the compound id is displayed in the form.
-- if the save is not successful, the user is notified of the error

**Comment by debboutr** [comment]
[Audio from ticketting...](https://drive.google.com/file/d/17s9t6EZiTi8lgf28g9J8IJ16RlRtXZ-8/view?usp=sharing)

**Comment by debboutr** [comment]
@cmgrulke there are 2 ways that a user may come across having a DefinedCompound in state with the editor

1. the user searches by CID or inchikey in the navbar w/ success
2. the user draws up an existing compound in the editor and it is found in the DB

In my interpretation, if #1 above is the case, and then edits are made, the user would PATCH the existing compound. But what is the workflow if they had arrived at a CID while drawing?? same thing?? Currently I have it setup where if they are drawing and a CID is found, the save button is disabled until they make edits, but in the case that they do edit from there should they be saving to the existing CID or a new one? Or is there something else that should happen?

**Comment by cmgrulke** [comment]
I accept the UI failing to successfully patch the molfile.  I would like to be able to see that the UI is trying to PATCH or POST based on the behavior described above, but I will be able to see that in dev tools.  The PATCHing error should be resolved by chemcurator_vuejs #73 as you mention above.

**Comment by michael-barbour** [comment]
@cmgrulke So I think #58 and this ticket might have a bug related to the perceived workflow.

If you want to add Water to the db as a compound, you would:

1. Click the Oxygen button
2. Click the canvas
3. Save

If you THEN wanted to add Hydrogen Peroxide, you would:

1. Click the Oxygen button
2. Click the canvas to add a H2O
3. Click the water molecule and drag
4. Save

The issue I see is how patching is meant to work here.  As soon as you click to add Oxygen to the canvas, the compound for water is loaded.  This matches acceptance criteria 2.

> if the defined compound exists, its CID is displayed in the CID field on the form (otherwise left blank)

the next edit changes the CID water to reflect Hydrogen Peroxide and the subsequent change changes the compound that used to be associated with Water to be instead associated with Hydrogen Peroxide.  This matches acceptance criteria 3

> if a defined compound does exist, but its structure exported from the ketcher window does not match with the one stored by the api, a "Save Structure" button is available and when the save structure button is pressed, the structure is PATCHED
> -- if the save is successful, the compound id is displayed in the form.
> -- if the save is not successful, the user is notified of the error

There is no way to add a Hydrogen Peroxide compound without first loading a different compound.

@debboutr solution fits within the acceptance criteria but seems to break expected functionality.  If we could set up a meeting sometime I'd love to be able to hash out some of the workflows of this page.  It's seems very complex and I'm not entirely sure this is the only part that I'm confused about.

**Comment by cmgrulke** [comment]
So, the patching to a CID occurs only in cases where the CID exists for the editted inchikey in the database already would remain unchanged.  So the case you have above of changing a water to a hydrogen peroxide would result in "Save molecule" eliciting a POST (if Hydrogen Peroxide was not already a stored compound in the database and therefore was not found when searching based on the edited structure).  If hydrogen peroxide was already in the database and the layout of the hydrogen peroxide you had was different than the hydrogen peroxide in the database (i.e. rotated), then "Save Structure" would issue a patch altering the molfile associated with that resource to store the new layout.

**Comment by debboutr** [comment]
@cmgrulke the issue we are describing here is concerned with an edit session being hijacked from a found compound. Let's say I want to draw up some exotic compound let's say acetylsalicylic acid, if the editor/user were to begin by drawing an oxygen atom in the window, that would create an H20 molecule, **if we had that in the DB, they would never be able to get off of editing the water compound that we have stored and onto the compound that they want to be editing, what will they do then?** Maybe this is something that we are missing between the whole compound=>substance connection that we don't know of, but it feels to me that if a compound is found a user might decide whether or not that is the intended compound that they are editing, else they may be stuck with whatever may come up while editing. Hope this makes sense, and you may already be aware of it, maybe the structure info will already be populated when making edits to a compound? that'd be news to me, but it would make sense that this might be the case. Can we expect that the user will rarely begin editing a compound from a blank editor?

**Comment by michael-barbour** [comment]
We talked to @cmgrulke and the general take away is that we made an assumption on when the molfile is going to be loaded into the editor window and what is considered an "edit".  

When the inchikeys match the CID stays the same.  If the inchikeys don't match the CID and compound info is cleared.

The above workflow would be like this:

1.     Click the Oxygen button
2.     Click the canvas to add a H2O
3.     Click the water molecule and drag (inchikeys don't match.  CID is cleared)
4.     Save (POST as no CID is loaded)
